plugins/janus_nosip.c

```diff
@@ -0,0 +1,2213 @@
+		janus_nosip_session *session = NULL;
+		janus_mutex_lock(&sessions_mutex);
+		if(g_hash_table_lookup(sessions, msg->handle) != NULL ) {
+			session = (janus_nosip_session *)msg->handle->plugin_handle;
+		}
+		janus_mutex_unlock(&sessions_mutex);
+		if(!session) {
+			JANUS_LOG(LOG_ERR, "No session associated with this handle...\n");
+			janus_nosip_message_free(msg);
+			continue;
+		}
+		if(session->destroyed) {
+			janus_nosip_message_free(msg);
+			continue;
+		}
+		/* Handle request */
+		error_code = 0;
+		root = msg->message;
+		if(msg->message == NULL) {
+			JANUS_LOG(LOG_ERR, "No message??\n");
+			error_code = JANUS_NOSIP_ERROR_NO_MESSAGE;
+			g_snprintf(error_cause, 512, "%s", "No message??");
+			goto error;
+		}
+		if(!json_is_object(root)) {
+			JANUS_LOG(LOG_ERR, "JSON error: not an object\n");
+			error_code = JANUS_NOSIP_ERROR_INVALID_JSON;
+			g_snprintf(error_cause, 512, "JSON error: not an object");
+			goto error;
+		}
+		JANUS_VALIDATE_JSON_OBJECT(root, request_parameters,
+			error_code, error_cause, TRUE,
+			JANUS_NOSIP_ERROR_MISSING_ELEMENT, JANUS_NOSIP_ERROR_INVALID_ELEMENT);
+		if(error_code != 0)
+			goto error;
+		json_t *request = json_object_get(root, "request");
+		const char *request_text = json_string_value(request);
+		json_t *result = NULL, *localjsep = NULL;
+
+		if(!strcasecmp(request_text, "generate") || !strcasecmp(request_text, "process")) {
+			/* Shared code for two different requests:	
+			 * 		generate: Take a JSEP offer or answer and generate a barebone SDP the application can use
+			 * 		process: Process a remote barebone SDP, and match it to the one we may have generated before */
+			gboolean generate = !strcasecmp(request_text, "generate") ? TRUE : FALSE;
+			if(generate) {
+				JANUS_VALIDATE_JSON_OBJECT(root, generate_parameters,
+					error_code, error_cause, TRUE,
+					JANUS_NOSIP_ERROR_MISSING_ELEMENT, JANUS_NOSIP_ERROR_INVALID_ELEMENT);
+			} else {
+				JANUS_VALIDATE_JSON_OBJECT(root, process_parameters,
+					error_code, error_cause, TRUE,
+					JANUS_NOSIP_ERROR_MISSING_ELEMENT, JANUS_NOSIP_ERROR_INVALID_ELEMENT);
+			}
+			if(error_code != 0)
+				goto error;
+			/* Any SDP to handle? if not, something's wrong */
+			const char *msg_sdp_type = json_string_value(json_object_get(generate ? msg->jsep : root, "type"));
+			const char *msg_sdp = json_string_value(json_object_get(generate ? msg->jsep : root, "sdp"));
+			if(!msg_sdp) {
+				JANUS_LOG(LOG_ERR, "Missing SDP\n");
+				error_code = JANUS_NOSIP_ERROR_MISSING_SDP;
+				g_snprintf(error_cause, 512, "Missing SDP");
+				goto error;
+			}
+			if(!msg_sdp_type || (strcasecmp(msg_sdp_type, "offer") && strcasecmp(msg_sdp_type, "answer"))) {
+				JANUS_LOG(LOG_ERR, "Missing or invalid SDP type\n");
+				error_code = JANUS_NOSIP_ERROR_MISSING_SDP;
+				g_snprintf(error_cause, 512, "Missing or invalid SDP type");
+				goto error;
+			}
+			gboolean offer = !strcasecmp(msg_sdp_type, "offer");
+			if(strstr(msg_sdp, "m=application")) {
```

Comment:
In case of an answer you're right, while I guess we'll still need tor reject it for offers. Will do that.